catalog/resource_grant.go

```diff
@@ -0,0 +1,187 @@
+		})
+	}
+	// STEP 2: non overlap - simply remove
+	for principal, remove := range remote {
+		_, ok := configured[principal]
+		if ok { // already handled in STEP 1
+			continue
+		}
+		diff.Changes = append(diff.Changes, permissions.UnityCatalogPermissionsChange{
+			Principal: principal,
+			Remove:    util.SetToSlice(remove),
+		})
+	}
+	// so that we can deterministic tests
+	sort.Slice(diff.Changes, func(i, j int) bool {
+		return diff.Changes[i].Principal < diff.Changes[j].Principal
+	})
+	return diff
+}
+
+// replacePermissionsForPrincipal merges removal diff of existing permissions on the platform
+func replacePermissionsForPrincipal(a permissions.UnityCatalogPermissionsAPI, securable string, name string, principal string, list permissions.UnityCatalogPermissionsList) error {
+	existing, err := a.GetPermissions(securable, name)
+	if err != nil {
+		return err
+	}
+	return a.UpdatePermissions(securable, name, diffPermissionsForPrincipal(principal, list, existing))
+}
+
+func filterPermissionsForPrincipal(in permissions.UnityCatalogPermissionsList, principal string) (out permissions.UnityCatalogPermissionsList) {
+
+	for _, v := range in.Assignments {
+		if v.Principal == principal {
+			out.Assignments = append(out.Assignments, v)
+		}
+	}
+	return
+}
+
+func ResourceGrant() *schema.Resource {
+	s := map[string]*schema.Schema{
+		"principal": {
+			Type:     schema.TypeString,
+			Required: true,
+			ForceNew: true,
+		},
+		"privileges": {
+			Type:     schema.TypeSet,
+			Required: true,
+			Elem:     &schema.Schema{Type: schema.TypeString},
+			Set:      schema.HashString,
+		},
+	}
+	// Handle all of the securable resource types
+	allFields := []string{}
+	for field := range permissions.Mappings {
+		allFields = append(allFields, field)
+	}
+
+	for _, field := range allFields {
+		s[field] = &schema.Schema{
+			Type:          schema.TypeString,
+			Optional:      true,
+			ForceNew:      true,
+			ConflictsWith: util.SliceWithoutString(allFields, field),
+		}
+	}
+
+	return common.Resource{
+		Schema: s,
+		CustomizeDiff: func(ctx context.Context, d *schema.ResourceDiff) error {
+			if d.Id() == "" {
+				// unfortunately we cannot do validation before dependent resources exist with tfsdkv2
+				return nil
+			}
+			var grants permissions.UnityCatalogPermissionsList
+			common.DiffToStructPointer(d, s, &grants)
+			return permissions.Mappings.Validate(d, grants)
+		},
+		Create: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			principal := d.Get("principal").(string)
+			privileges := util.ToStringSlice(d.Get("privileges").(*schema.Set).List())
+			var grants = permissions.UnityCatalogPermissionsList{
+				Assignments: []permissions.UnityCatalogPrivilegeAssignment{
+					{
+						Principal:  principal,
+						Privileges: privileges,
+					},
+				},
+			}
+			securable, name := permissions.Mappings.KeyValue(d)
+			unityCatalogPermissionsAPI := permissions.NewUnityCatalogPermissionsAPI(ctx, c)
```

Comment:
for the new resource, we should use go sdk to avoid tech debt. `databricks_grants` can be refactored later

Reply:
I'm not sure how I can address this in the scope of this PR. The `common.Resource` struct requires `common.DatabricksClient`

Reply:
so you can initialise the sdk client using

```go
w, err := c.WorkspaceClient()
if err != nil {
	return err
}

w.Grants.Update(ctx, catalog.UpdatePermissions{
	Changes:       permissionsChange,
	SecurableType: securable,
	FullName:      name,
})
```

Reply:
I made a start (see https://github.com/siteminder-au/terraform-provider-databricks/compare/resource_grant...siteminder-au:terraform-provider-databricks:sdk-update), but I have some concerns that I don't know enough about the backend to make this change. For example, the permissions for shares changes URL & using the new types from the sdk, the StructToSchema generates schema with different Required & Optional settings.

Using the existing `resource_grants.go` as a base, I was comfortable to make the change as it is amounted to little more than verifying that the content of the HTTP requests match what is expected.

I can proceed if you need me to, but I will need some direction

Reply:
See the diff above for the `resource_grant_test.go` for the share related tests

Reply:
Re the different Optional & Required behaviour, I narrowed it down to the difference of `omitempty` in jsontag
e.g. existing
```go
// PrivilegeAssignment reflects on `grant` block
type PrivilegeAssignment struct {
	Principal  string   `json:"principal"`
	Privileges []string `json:"privileges" tf:"slice_set"`
}
```

sdk
```go
type PrivilegeAssignment struct {
	// The principal (user email address or group name).
	Principal string `json:"principal,omitempty"`
	// The privileges assigned to the principal.
	Privileges []Privilege `json:"privileges,omitempty"`

	ForceSendFields []string `json:"-"`
}
```

Reply:
sorry I should have been more specific, for `StructToSchema`, we can leave it to use the same custom struct as `resource_grants` - there are some outstanding work on how we can add additional tf annotation to those structs

it should be easier to swap these API calls
```go
unityCatalogPermissionsAPI := permissions.NewUnityCatalogGrantsAPI(ctx, c)
err := replacePermissionsForPrincipal(unityCatalogPermissionsAPI, securable, name, principal, grants)
```
with the appropriate sdk call (need adjustment for type), e.g.
```go
existing, err := w.Grants.GetBySecurableTypeAndFullName(ctx, catalog.SecurableType(securable), name)
if err != nil {
  return err
}

w.Grants.Update(ctx, catalog.UpdatePermissions{
  Changes:       diffPermissionsForPrincipal(principal, list, existing),
  SecurableType: securable,
  FullName:      name,
})
```

the two resources with special handling here are `shares` (different API call `w.Shares.UpdatePermissions(ctx, sharing.UpdateSharePermissions{})`) and `connection` (reserved TF keyword)

Reply:
Should I revert the changes to `resource_grants` as the code diverges now ?

Reply:
@nkvuong I reverted those `resource_grants` changes & refactored `resource_grant` to use the SDK. I kept the `UnityCatalogPermissionsAPI` to hide the difference between Shares & Grants APIs.

Reply:
I can bring back the refactor easily enough also & update to use sdk, let me know

Reply:
If needed: https://github.com/siteminder-au/terraform-provider-databricks/compare/resource_grant...siteminder-au:terraform-provider-databricks:resource_grants_refactor

Reply:
you can raise another PR for the refactoring of `resource_grants` to Go SDK - makes it easier to review but also revert (if needed)